mypy/checkexpr.py

```diff
@@ -204,7 +204,13 @@ def visit_call_expr(self, e: CallExpr, allow_none_return: bool = False) -> Type:
                         or isinstance(typ, NameExpr) and node and node.kind == nodes.TYPE_ALIAS):
                     self.msg.type_arguments_not_allowed(e)
         self.try_infer_partial_type(e)
-        callee_type = self.accept(e.callee, always_allow_any=True)
+        if isinstance(e.callee, LambdaExpr):
+            type_context = CallableType([self.accept(a) for a in e.args], e.arg_kinds, e.arg_names,
```

Comment:
You need to take the argument kinds and names from the lambda expression (not the call expression) and map the actual argument to formal arguments. Currently this doesn't work correctly, for example:

```py
(lambda s, i: s[i])(i=0, s='x')
```